go-controller/pkg/libovsdbops/db_object_ids.go

```diff
@@ -0,0 +1,295 @@
+	id := objectIDs.OwnerControllerName + ":" + string(objectIDs.IDsType.ownerObjectType)
+	for _, key := range objectIDs.IDsType.GetIDs() {
+		id += ":" + objectIDs.ObjectIDs[key]
+	}
+	return id
+}
+
+// getExternalIDsWithoutPrimary returns the same ids as GetExternalIDs, but without PrimaryIDKey
+func (objectIDs *DbObjectIDs) getExternalIDsWithoutPrimary(omitEmpty bool) map[string]string {
+	externalIDs := map[string]string{
+		OwnerControllerKey: objectIDs.OwnerControllerName,
+		OwnerTypeKey:       string(objectIDs.IDsType.ownerObjectType),
+	}
+	for _, key := range objectIDs.IDsType.GetIDs() {
+		if omitEmpty && objectIDs.ObjectIDs[key] == "" {
+			continue
+		}
+		externalIDs[key] = objectIDs.ObjectIDs[key]
+	}
+	return externalIDs
+}
+
+// GetDbObjectIDsFromExternalIDs is used to parse object ExternalIDs, it sets DbObjectIDs.OwnerControllerName based
+// on OwnerControllerKey key, and verifies OwnerControllerKey value matches given objectIDsType.
+// All the other ids (except for PrimaryIDKey) will be set to DbObjectIDs.ObjectIDs, even keys that are not a part of objectIDsType.
+func GetDbObjectIDsFromExternalIDs(objectIDsType ObjectIDsType, externalIDs map[string]string) (*DbObjectIDs, error) {
+	if externalIDs[OwnerTypeKey] != string(objectIDsType.ownerObjectType) {
+		return nil, fmt.Errorf("expected ExternalID %s to equal %s, got %s",
+			OwnerTypeKey, string(objectIDsType.ownerObjectType), externalIDs[OwnerTypeKey])
+	}
+	if externalIDs[OwnerControllerKey] == "" {
+		return nil, fmt.Errorf("required ExternalID %s is empty", OwnerControllerKey)
+	}
+	objIDs := map[string]string{}
+	for key, value := range externalIDs {
+		if key != OwnerTypeKey && key != OwnerControllerKey && key != PrimaryIDKey {
+			objIDs[key] = value
+		}
+	}
+	return NewDbObjectIDs(objectIDsType, externalIDs[OwnerControllerKey], objIDs), nil
+}
+
+// hasExternalIDs interface should only include types that use new ExternalIDs from DbObjectIDs.
+type hasExternalIDs interface {
+	GetExternalIDs() map[string]string
+}
+
+// GetNoOwnerPredicate should only be used on initial sync when switching to new ExternalIDs.
+// Otherwise, use GetPredicate with the specific OwnerControllerKey id.
+func GetNoOwnerPredicate[T hasExternalIDs]() func(item T) bool {
+	return func(item T) bool {
+		return item.GetExternalIDs()[OwnerControllerKey] == ""
+	}
+}
+
+// GetPredicate returns a predicate to search for db obj of type T (a set of supported types is defined in hasExternalIDs).
+// Only non-empty ids will be matched (that always includes DbObjectIDs.OwnerTypeKey and DbObjectIDs.OwnerControllerName),
+// but
+// Other IDs may be empty and will be ignored in the filtering, additional filter function f may be passed, or set to nil.
+func GetPredicate[nbdbT hasExternalIDs](dbIndex *DbObjectIDs, f func(item nbdbT) bool) func(item nbdbT) bool {
+	return func(item nbdbT) bool {
+		dbExternalIDs := item.GetExternalIDs()
+		indexIDs := dbIndex.getExternalIDsWithoutPrimary(true)
```

Comment:
I guess that using a predicate does not preclude uniquely identifying an object. In this regard, why would we not use the unique id primarily if available?

Reply:
yeah we can do that with an additional `getUniqueIDWithCheck` function